x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskRandomTests.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ccr.action;
+
+import org.elasticsearch.action.UnavailableShardsException;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.index.seqno.LocalCheckpointTracker;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.translog.Translog;
+import org.elasticsearch.test.ESTestCase;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.LongConsumer;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class ShardFollowNodeTaskRandomTests extends ESTestCase {
+
+    public void testSingleReaderWriter() throws Exception {
+        TestRun testRun = createTestRun(randomNonNegativeLong(), randomNonNegativeLong(), randomIntBetween(1, 2048));
+        ShardFollowNodeTask task = createShardFollowTask(1, testRun);
+        startAndAssertAndStopTask(task, testRun);
+    }
+
+    public void testMultipleReaderWriter() throws Exception {
+        int concurrency = randomIntBetween(2, 8);
+        TestRun testRun = createTestRun(0, 0, 1024);
+        ShardFollowNodeTask task = createShardFollowTask(concurrency, testRun);
+        startAndAssertAndStopTask(task, testRun);
+    }
+
+    private void startAndAssertAndStopTask(ShardFollowNodeTask task, TestRun testRun) throws Exception {
+        task.start(testRun.startSeqNo - 1, testRun.startSeqNo - 1);
+        assertBusy(() -> {
+            ShardFollowNodeTask.Status status = task.getStatus();
+            assertThat(status.getLeaderGlobalCheckpoint(), equalTo(testRun.finalExpectedGlobalCheckpoint));
+            assertThat(status.getFollowerGlobalCheckpoint(), equalTo(testRun.finalExpectedGlobalCheckpoint));
+            assertThat(status.getIndexMetadataVersion(), equalTo(testRun.finalIndexMetaDataVerion));
+        });
+
+        task.markAsCompleted();
```

Comment:
Can we only cancel the task randomly?

Reply:
Cancelling is the only way to stop the shard follow task, so I don't think that this is possible, otherwise things may still be happening in the background.

Reply:
The timeout is now hardcoded 10ms, which I think is ok. I will add an assert to ensure when it gets changed, the dev who will change it know why we like to keep this timeout low.